keps/672-execution-policy/README.md

```diff
@@ -0,0 +1,421 @@
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+    - name: trainer-node
+      template:
+        spec:
+          parallelism: 3
+          completions: 3
+          template:
+            spec:
+              containers:
+                - name: trainer
+                  image: docker.io/kubeflow/deepspeed-trainer
+                  resources:
+                    limits:
+                      nvidia.com/gpu: 5
+                  volumeMounts:
+                    - mountPath: /workspace/dataset
+                      name: dataset-initializer
+                    - mountPath: /workspace/pre-trained-model
+                      name: model-initializer
+              volumes:
+                - name: dataset-initializer
+                  persistentVolumeClaim:
+                    claimName: dataset-initializer
+                - name: model-initializer
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+```
+
+#### Story 3
+
+As an HPC user, I want to run a set of simulations with MPI that complete successfully before
+using the data for a next step analysis.
+
+As an HPC user I want to scale up my simulation job only after a small set have completed
+successfully. For that case, I could imagine essentially the same replicatedJob done twice,
+just with a larger size. If the outcome of the first isn't success you wouldn't launch
+the larger bulk of work.
+
+### Risks and Mitigations
+
+This API will not allow to describe DAGs to avoid workflow manager features in JobSet.
+The goal is to only focus on Job sequence to cover model training/HPC use-cases.
+
+## Design Details
+
+### API Details
+
+```golang
+type JobSetSpec struct {
+	ExecutionPolicy *ExecutionPolicy `json:"executionPolicy,omitempty"`
+}
+
+type ExecutionPolicyOption string
+
+const (
+	// This is the default settings.
+	// AnyOrder means that Jobs will be started in any order.
+	AnyOrder ExecutionPolicyOption = "AnyOrder"
+
+	// InOrder starts the ReplicatedJobs in order that they are listed. Jobs within a ReplicatedJob
+	// will still start in any order.
+	InOrder ExecutionPolicyOption = "InOrder"
+)
+
+type ExecutionPolicy struct {
+	// Order in which Jobs will be created.
+	ExecutionPolicyOrder ExecutionPolicyOption `json:"executionPolicyOrder"`
+
+	// After all ReplicatedJobs reach this status, the JobSet will create the next ReplicatedJobs.
+	ExecutionPolicyRule []ExecutionPolicyRule `json:"rules"`
+}
+
+// ExecutionPolicyRule represents the execution policy rule for Job sequence.
+type ExecutionPolicyRule struct {
+
+	// Names of the replicated Jobs that applied the status.
+	TargetReplicatedJobs []string `json:"targetReplicatedJobs"`
+
+	// Status the target ReplicatedJobs must reach before subsequent ReplicatedJobs begin executing.
+	WaitForReplicatedJobsStatus ReplicatedJobsStatusOption `json:"waitForReplicatedJobsStatus"`
+}
+
+type ReplicatedJobsStatusOption string
```

Comment:
Sure, let me add it.
I am wondering, do we want to introduce any new JobSet condition to improve UX ?
Similar to what we've done with StartupPolicy: https://github.com/kubernetes-sigs/jobset/blob/main/pkg/controllers/jobset_controller.go#L511

Reply:
I think, for ExecutionPolicy we might want to introduce another condition. For the StartupPolicy, currently user can see the following:
```yaml
  status:
    conditions:
    - lastTransitionTime: "2024-10-07T14:55:37Z"
      message: in order startup policy is in progress
      reason: InOrderStartupPolicyInProgress
      status: "False"
      type: StartupPolicyInProgress
    - lastTransitionTime: "2024-10-07T14:55:38Z"
      message: in order startup policy has completed
      reason: InOrderStartupPolicyCompleted
      status: "True"
      type: StartupPolicyCompleted
    - lastTransitionTime: "2024-10-07T14:55:38Z"
      message: jobset completed successfully
      reason: AllJobsCompleted
      status: "True"
      type: Completed
```

I am not sure if we should introduce another condition which shows that ReplicatedJob-X is completed.

Reply:
> I think, for ExecutionPolicy we might want to introduce another condition. For the StartupPolicy, currently user can see the following:
> 
> ```yaml
>   status:
>     conditions:
>     - lastTransitionTime: "2024-10-07T14:55:37Z"
>       message: in order startup policy is in progress
>       reason: InOrderStartupPolicyInProgress
>       status: "False"
>       type: StartupPolicyInProgress
>     - lastTransitionTime: "2024-10-07T14:55:38Z"
>       message: in order startup policy has completed
>       reason: InOrderStartupPolicyCompleted
>       status: "True"
>       type: StartupPolicyCompleted
>     - lastTransitionTime: "2024-10-07T14:55:38Z"
>       message: jobset completed successfully
>       reason: AllJobsCompleted
>       status: "True"
>       type: Completed
> ```
> 
> I am not sure if we should introduce another condition which shows that ReplicatedJob-X is completed.

Hmm, thinking about it more I think it would be simpler to extend StartupPolicy to support this behavior rather than create a new ExecutionPolicy API. All of the conditions would be reusable, and the code changes would be more minimal I believe.

Reply:
>All of the conditions would be reusable, and the code changes would be more minimal I believe.

I think, the status should look a little bit different.
Do we want to introduce the same condition if user sets sequence of ReplicatedJobs ?

Reply:
I think I tried this but we run into a problem with conditions. Conditions cannot be duplicated (either status is True or False) and adding particular "messages" to the StartupPolicy condition will cause an error.

Reply:
Yes, that is correct, they should be unique. I am not sure if adding that condition gives a lot of value for the user:
`type: StartupPolicyInProgress, status: True` or `type: ExecutionPolicyInProgress, status: True`

Reply:
If we extend StartupPolicy rather than adding a new ExecutionPolicy API, we should be able to continue using the same condition (e.g., `type: StartupPolicyInProgress, status: True`) regardless of whether the startupPolicy defined by the user waits for replicatedJob status Ready or Succeeded. Let me know if I'm misunderstanding the point you're discussing.

Reply:
Sure, that makes sense.